### PR TITLE
Wait for editor changes to be debounced before closing modal

### LINKED ISSUE #38020: Iframe Editor: incomplete content on modal close
When using the modal editor in the new product blocks editor, content doesn't always get completely saved and shown in the description state.

This seems like it may be partially due to our debouncing the `onChange` ( https://github.com/woocommerce/woocommerce/blob/66a10ee8ce821cab178dd6b8fa951135bd10bd22/packages/js/product-editor/src/components/iframe-editor/iframe-editor.tsx#L69-L76 ) but also because the Gutenberg `onChange` handler appears to be debounced as well.

Adding in buttons to "Save/Update" the description may also help remedy this by allowing more time for the debounce to complete prior to closing the modal.

### Screenshots

Image

### Reproducing

1. Enable the product blocks experience
2. Navigate to "Add product"
3. Click "Add description"
4. Add some content and quickly close the modal
5. Note that the content is not always complete

### Acceptance criteria

* [ ] The content should show in its entirety when the modal editor is closed

### PR DESCRIPTION
**Pull request by joshuatf**
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Making changes within the modal editor and quickly closing the editor previously would result in some of the changes being missed.

This PR capture the changes immediately, debounces on the onchange handler, and only closes the modal once the debouncing is complete.

Screen Shot 2023-05-11 at 12 16 05 PM

Closes #38020  .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Navigate to Tools -> WCA Test Helper -> Features and enable the new product blocks editing experience
2. Navigate to Products -> Add new
3. Click on the "Add description" button
4. Make changes within the editor and quickly close the modal
5. Note that all changes are reflected in the preview (or reopen the modal to confirm)
6. Open and close the editor without changes to make sure it closes without issue

### CODE REVIEW

File: packages/js/product-editor/src/components/modal-editor/modal-editor.tsx [hunk 1]
```diff
@@ -28,16 +29,40 @@ export function ModalEditor( {
...
 			className="woocommerce-modal-editor"
 			title={ title }
-			onRequestClose={ onClose }
+			onRequestClose={ handleClose }
 			shouldCloseOnClickOutside={ false }
 		>
 			<IframeEditor
 				initialBlocks={ initialBlocks }
-				onChange={ onChange }
+				onInput={ handleInput }
```
Comment (louwie17) [inline packages/js/product-editor/src/components/modal-editor/modal-editor.tsx]: It actually looks like you need both `onChange` and `onInput` :/ they are both being called for different circumstances.
For example, `onChange` only gets called when new blocks are created, but `onInput` when something within the block is updated. You can pretty easily see this if you console log both functions.
Also see this example: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/README.md?plain=1#L34-L35
And the description of the outputs: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/provider/README.md?plain=1#L14-L32

File: packages/js/product-editor/src/components/modal-editor/modal-editor.tsx [hunk 2]
```diff
@@ -28,16 +29,40 @@ export function ModalEditor( {
 	onClose,
 	title,
 }: ModalEditorProps ) {
+	const [ isDebouncing, setIsDebouncing ] = useState( false );
+	const [ isClosing, setIsClosing ] = useState( false );
+
+	useEffect( () => {
+		// Only close the modal once editor changes have persisted after debouncing.
+		if ( isClosing && ! isDebouncing ) {
+			onClose();
```
Comment (louwie17) [inline packages/js/product-editor/src/components/modal-editor/modal-editor.tsx]: This is a rather hacky way of doing this, and at-least at one point I managed to trigger the `onClose` but it wasn't being triggered until I made another block change (probably related to the difference between `onInput` and `onChange`.
But there is a much easier way to handle this, the `useDebounce` hook has a `flush` method that returns the last block result. So you can just do this within the `handleClose`:
```
const blocks = debouncedOnChange.flush();
if ( blocks ) {
  onChange( blocks );
}
onClose();
```
This way you can remove both the `isClosing` and `isDebouncing` states :)

**Review by louwie17 — CHANGES_REQUESTED** [review]
Thanks for the updates, I did leave some suggestions to do this differently, basically two things:
- Call `debouncedOnChange` for both `onInput` & `onChange` (this is actually necessary :/ see inline comment for more info)
- Make use of the `flush()` method that `useDebounce` returns.

**Review by louwie17 — CHANGES_REQUESTED** [review]
Thanks for the updates, I did leave some suggestions to do this differently, basically two things:
- Call `debouncedOnChange` for both `onInput` & `onChange` (this is actually necessary :/ see inline comment for more info)
- Make use of the `flush()` method that `useDebounce` returns.

File: packages/js/product-editor/src/components/modal-editor/modal-editor.tsx (hunk 1 above)
Comment (joshuatf) [reply packages/js/product-editor/src/components/modal-editor/modal-editor.tsx]: Good call.  I consoled the outputs and noticed this as well.

>A callback invoked when the blocks have been modified in a non-persistent manner.

I had read the above and re-reading it I'm still convinced the documentation is semantically incorrect.  I would expect the non-persistent callback (`onInput`) to always fire while the persistent callback (`onChange`) would fire when those changes were persisted.

This behavior is all-around a little odd and does not match the [browser API](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/input_event) as stated in the docs.  I'll open an issue upstream to get this fixed.

File: packages/js/product-editor/src/components/modal-editor/modal-editor.tsx (hunk 2 above)
Comment (joshuatf) [reply packages/js/product-editor/src/components/modal-editor/modal-editor.tsx]: >This is a rather hacky way of doing this, and at-least at one point I managed to trigger the onClose but it wasn't being triggered until I made another block change (probably related to the difference between onInput and onChange.

I suspect you managed to do this based on the buggy behavior of `onChange` and `onInput` discussed above.

>useDebounce hook has a flush method that returns the last block result

TIL!  Thanks for sharing this- this is a much easier way to implement this.

**Review by louwie17 — APPROVED** [review]
Thanks for the updates, this tested well 🚀

**Pull request merged**